File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Database/DatabaseObject.cs
// <copyright file="RemoteDatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Collections.Generic;
    using Meta;
    using System.Linq;
    using Allors.Protocol.Json.Api.Sync;

    internal class DatabaseObject
    {
        private static readonly HashSet<long> EmptySet = new HashSet<long>();

        private Permission[] deniedPermissions;
        private AccessControl[] accessControls;

        private Dictionary<IRelationType, object> roleByRelationType;
        private SyncResponseRole[] syncResponseRoles;

        internal DatabaseObject(Database database, long identity, IClass @class)
        {
            this.Database = database;
            this.Identity = identity;
            this.Class = @class;
            this.Version = 0;
        }

        internal DatabaseObject(Database database, ResponseContext ctx, SyncResponseObject syncResponseObject)
        {
            this.Database = database;
            this.Identity = syncResponseObject.Id;
            this.Class = (IClass)this.Database.MetaPopulation.FindByTag(syncResponseObject.ObjectType);
            this.Version = syncResponseObject.Version;
            this.syncResponseRoles = syncResponseObject.Roles;
            this.AccessControlIds = syncResponseObject.AccessControls != null ? new HashSet<long>(ctx.CheckForMissingAccessControls(syncResponseObject.AccessControls)) : EmptySet;
            this.DeniedPermissionIds = syncResponseObject.DeniedPermissions != null ? new HashSet<long>(ctx.CheckForMissingPermissions(syncResponseObject.DeniedPermissions)): EmptySet;
        }

        internal Database Database { get; }

        internal IClass Class { get; }

        internal long Identity { get; }

        internal long Version { get; private set; }

        internal HashSet<long> AccessControlIds { get; }

        internal HashSet<long> DeniedPermissionIds { get; private set; }

        private Dictionary<IRelationType, object> RoleByRelationType
        {
            get
            {
                if (this.syncResponseRoles != null)
                {
                    var meta = this.Database.MetaPopulation;

                    var metaPopulation = this.Database.MetaPopulation;
                    this.roleByRelationType = this.syncResponseRoles.ToDictionary(
                        v => (IRelationType)meta.FindByTag(v.RoleType),
                        v =>
                        {
                            var roleType = ((IRelationType)metaPopulation.FindByTag(v.RoleType)).RoleType;

                            var objectType = roleType.ObjectType;
                            if (objectType.IsUnit)
                            {
                                return UnitConvert.FromString(roleType.ObjectType.Tag, v.Value);
                            }

                            if (roleType.IsOne)
                            {
                                return v.Object;
                            }

                            return v.Collection;
                        });

                    this.syncResponseRoles = null;
                }

                return this.roleByRelationType;
            }
        }

        private AccessControl[] AccessControls =>
            this.accessControls = this.accessControls switch
            {
                null when this.AccessControlIds == null => Array.Empty<AccessControl>(),
                null => this.AccessControlIds
                    .Select(v => this.Database.AccessControlById[v])
                    .ToArray(),
                _ => this.accessControls
            };

        private Permission[] DeniedPermissions =>
            this.deniedPermissions = this.deniedPermissions switch
            {
                null when this.DeniedPermissionIds == null => Array.Empty<Permission>(),
                null => this.DeniedPermissionIds
                    .Select(v => this.Database.PermissionById[v])
                    .ToArray(),
                _ => this.deniedPermissions
            };

        public object GetRole(IRoleType roleType)
        {
            object @object = null;
            this.RoleByRelationType?.TryGetValue(roleType.RelationType, out @object);
            return @object;
        }

        public bool IsPermitted(Permission permission) =>
            permission != null &&
            !this.DeniedPermissions.Contains(permission) &&
            this.AccessControls.Any(v => v.PermissionIds.Any(w => w == permission.Id));

        internal void UpdateDeniedPermissions(long[] deniedPermissions)
        {
            if (this.deniedPermissions == null)
            {
                this.DeniedPermissionIds = EmptySet;
            }
            else
            {
                if (!this.DeniedPermissionIds.SetEquals(deniedPermissions))
                {
                    this.DeniedPermissionIds = new HashSet<long>(deniedPermissions);
                }
            }
        }
    }
}

File: System/Workspace/Typescript/libs/workspace/domain/json/system/src/lib/Database/DatabaseObject.ts
import { SyncResponseObject, SyncResponseRole } from '@allors/protocol/json/system';
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { Database } from './Database';
import { AccessControl } from './Security/AccessControl';
import { Permission } from './Security/Permission';
import { ResponseContext } from './Security/ResponseContext';

export class DatabaseObject {
  version: number;

  private _roleByRelationType: Map<RelationType, unknown>;
  private _syncResponseRoles: SyncResponseRole[];

  private _accessControls: AccessControl[];
  private _accessControlIds: number[];

  private _deniedPermissions: Permission[];
  private _deniedPermissionIds: number[];

  constructor(public readonly database: Database, public readonly identity: number, public readonly cls: Class) {
    this.version = 0;
  }

  static fromResponse(database: Database, ctx: ResponseContext, syncResponseObject: SyncResponseObject): DatabaseObject {
    const obj = new DatabaseObject(database, syncResponseObject.i, database.metaPopulation.metaObjectByTag.get(syncResponseObject.t) as Class);
    obj.version = syncResponseObject.v;
    obj._syncResponseRoles = syncResponseObject.r;
    obj._accessControlIds = ctx.checkForMissingAccessControls(syncResponseObject.a);
    obj._deniedPermissionIds = ctx.checkForMissingPermissions(syncResponseObject.d);
    return obj;
  }

  get roleByRelationType(): Map<RelationType, unknown> {
    if (this._syncResponseRoles != null) {
      const meta = this.database.metaPopulation;
      this._roleByRelationType = new Map(
        this._syncResponseRoles.map((v) => {
          const relationType = meta.metaObjectByTag.get(v.t) as RelationType;
          const roleType = relationType.roleType;
          const objectType = roleType.objectType;
          let role: unknown;

          if (objectType.isUnit) {
            role = v.v;
            // TODO:
            // return UnitConvert.FromString(roleType.ObjectType.Tag, v.Value);
          } else {
            if (roleType.isOne) {
              role = v.o;
            } else {
              role = v.c;
            }
          }

          return [relationType, role];
        })
      );

      this._syncResponseRoles = null;
    }

    return this._roleByRelationType;
  }

  get accessControls(): AccessControl[] {
    return (this._accessControls ??= this._accessControlIds?.map((v) => this.database.accessControlById.get(v)) ?? []);
  }

  get deniedPermissions(): Permission[] {
    return (this._deniedPermissions ??= this._deniedPermissionIds?.map((v) => this.database.permissionById.get(v)) ?? []);
  }

  getRole(roleType: RoleType): unknown {
    return this.roleByRelationType.get(roleType.relationType);
  }

  isPermitted(permission: Permission): boolean {
    return !!permission && !this.deniedPermissions.includes(permission) && !!this.accessControls.filter((v) => v.permissionIds.has(permission.id)).length;
  }

  updateDeniedPermissions(updatedDeniedPermissions: number[]) {
    this._deniedPermissionIds = updatedDeniedPermissions;
  }
}
